File: example/xx_three/clustered.cpp
#include <xx_three/xx_three.h>
#include <gfx-pbr/Api.h>
#include <gfx-obj/Api.h>
#include <ecs/ECS.hpp>

#include <stl/vector.hpp>

#define CLUSTERED 1

EX(xx_clustered)
{
#if UI
	UNUSED(dockbar);
	SceneViewer& viewer = ui::scene_viewer(parent);
	Scene& scene = viewer.m_scene;
#else
	static Scene scene = Scene(app.m_gfx);
	static GfxViewer viewer = GfxViewer(window, scene);
#endif

	static ImporterOBJ obj_importer(app.m_gfx);

#if UI
	ui::orbit_controls(viewer);
	//controls.minDistance = 120;
	//controls.maxDistance = 320;
#endif

	//var bloom = new THREE.UnrealBloomPass(new THREE.Vector2(), 0.8, 0.6, 0.8);
	//bloom.renderToScreen = true;

#if CLUSTERED
	Camera& camera = viewer.m_camera;
	viewer.m_viewport.set_clustered(app.m_gfx);
#endif

	constexpr float radius = 75.f;

	struct ExLight
	{
		Node3* parent;
		Node3* node;
		Colour color;
		float radius = radius;
		float decay = 1.f;
		float sy, sr, sc;
		float py, pr, pc;
		float dir;
	};

	static vector<ExLight> lights = {};

	if(init)
	{
		Camera& camera = viewer.m_camera;
		camera.m_fov = 40.f; camera.m_near = 1.f; camera.m_far = 2000.f;
		camera.m_eye = vec3(0.f, 0.f, 240.f);

		Zone& env = scene.m_env;
		env.m_radiance.m_ambient = rgb(0xffffff) * 0.33f;

		env.m_background.m_colour = rgb(0x111111);
		viewer.m_viewport.m_clear_colour = rgb(0x111111);

		Model& model = *app.m_gfx.models().file("WaltHead");
		
		Model& sphere = app.m_gfx.shape(Sphere(0.5f));
		Model& big_sphere = app.m_gfx.shape(Sphere(0.5f * 6.66f));

		auto create_material = [&](const string& name, auto init) -> Material* { return &app.m_gfx.materials().create(name, init); };

		Program& solid = app.m_gfx.programs().fetch("solid");
		Program& phong = app.m_gfx.programs().fetch("pbr/phong");
		Program& three = app.m_gfx.programs().fetch("pbr/three");

		Material* materials[] = {
			create_material("first",  [&](Material& m) { m.m_program = &three; m.m_pbr.m_albedo = rgb(0x888888); m.m_pbr.m_metallic = 1.0f; m.m_pbr.m_roughness = 0.66f; }),
			create_material("second", [&](Material& m) { m.m_program = &three; m.m_pbr.m_albedo = rgb(0x666666); m.m_pbr.m_metallic = 0.1f; m.m_pbr.m_roughness = 0.33f; }),
			create_material("third",  [&](Material& m) { m.m_program = &phong; m.m_phong.m_diffuse = rgb(0x777777); m.m_phong.m_shininess = 20.f; }),
			create_material("fourth", [&](Material& m) { m.m_program = &phong; m.m_phong.m_diffuse = rgb(0x555555); m.m_phong.m_shininess = 10.f; m.m_phong.m_toon = true; }),
		};

		size_t transparent = randi(0, 3);
		materials[transparent]->m_alpha.m_alpha = 0.9f;

		//transparent : tIndex == = index ? true : false,
		//mtl.uniforms["opacity"].value = tIndex == = index ? 0.9 : 1;
		//mtl.side = tIndex == = index ? THREE.FrontSide : THREE.DoubleSide;

		size_t index = 0;
		for(Material* material : materials)
		{
			const float i = float(index++);
			const vec3 position = vec3(sin(i * c_pi2) * radius, 0.f, cos(i * c_pi2) * radius);
			const quat rotation = quat(vec3(0.f, i * c_pi2, 0.f));

			Node3& n = gfx::nodes(scene).add(Node3(position, rotation));

			Node3& m = gfx::nodes(scene).add(Node3(n.m_transform * bxTRS(vec3(1.f), ZeroQuat, vec3(0.f, -37.f, 0.f))));
			Item& it = gfx::items(scene).add(Item(m, model, 0U, material));
			UNUSED(it);


			for(size_t i = 0; i < 8; i++)
			{
				const Colour color = hsl(randf(), 1.0, 0.5f);

				Material& ml = app.m_gfx.materials().create("light" + material->m_name + to_string(i), [&](Material& m) { 
					m.m_program = &solid; m.m_solid.m_colour = color; 
				});

				Material& mla = app.m_gfx.materials().create("lightalpha" + material->m_name + to_string(i), [&](Material& m) {
					m.m_program = &solid; m.m_solid.m_colour = color; m.m_alpha.m_alpha = 0.033f;
				});

				Node3& l = gfx::nodes(scene).add(Node3());
				Item& i0 = gfx::items(scene).add(Item(l, sphere, 0U, &ml)); // MaterialSolid(color)));
			
				//Item& i1 = gfx::items(scene).add(Item(l, big_sphere, 0U, &ma)); // MaterialSolid(color), MaterialAlpha(0.033f));
				//l.children[1].scale.set(6.66, 6.66, 6.66);

				Light& light = gfx::lights(scene).add(Light(l, LightType::Point, false, color, 1.f, radius));
				light.m_attenuation = 1.f;

				lights.push_back({
					&n,
					&l,
					color,
					radius,
					1.f,
					randf(),
					randf(),
					randf(),
					randf() * c_pi,
					randf() * c_pi,
					randf() * c_pi,
					randf() > 0.5 ? 1.f : -1.f
				});
			}
		}
	}

	float time = app.m_gfx.m_time;

	for(ExLight& l : lights)
	{
		float r = 0.8f + 0.2f * sin(l.pr + (0.6f + 0.3f * l.sr) * time);
		float x = (sin(l.pc + (0.8f + 0.2f * l.sc) * time * l.dir)) * r * radius;
		float z = (cos(l.pc + (0.8f + 0.2f * l.sc) * time * l.dir)) * r * radius;
		float y = sin(l.py + (0.8f + 0.2f * l.sy) * time) * r * 32.f;
		l.node->derive(*l.parent, vec3(x, y, z));
	}
}

File: docs/editor/xx_three/clustered.js
// clustered.js

var viewer = two.ui.scene_viewer(panel);
two.ui.orbit_controls(viewer);

var radius = 75.0;

var viewport = viewer.viewport;

if (!viewport.clustered) {
    // @todo fix clustered
    //viewport.set_clustered(app.gfx);
}

if (init) {
    this.importerOBJ = new two.ImporterOBJ(app.gfx);

    var camera = viewer.camera;
    camera.fov = 40.0; camera.near = 1.0; camera.far = 2000.0;
    camera.eye = new two.vec3(0.0, 0.0, 240.0);
    
    var scene = viewer.scene;

    var env = scene.env;
    env.radiance.ambient = new two.Colour(0.33);

    env.background.colour = two.rgb(0x111111);
    viewport.clear_colour = two.rgb(0x111111);
    
    var solid = app.gfx.programs.file('solid');
    var phong = app.gfx.programs.file('pbr/phong');
    var three = app.gfx.programs.file('pbr/three');

    var m0 = app.gfx.materials.create('m0');
    var m1 = app.gfx.materials.create('m1');
    var m2 = app.gfx.materials.create('m2');
    var m3 = app.gfx.materials.create('m3');
    
    m0.program = three; m0.pbr.albedo.value = two.rgb(0x888888); m0.pbr.metallic.value = 1.0; m0.pbr.roughness.value = 0.66;
    m1.program = three; m1.pbr.albedo.value = two.rgb(0x666666); m1.pbr.metallic.value = 0.1; m1.pbr.roughness.value = 0.33;
    m2.program = phong; m2.phong.diffuse.value = two.rgb(0x777777); m2.phong.shininess.value = 20;
    m3.program = phong; m3.phong.diffuse.value = two.rgb(0x555555); m3.phong.shininess.value = 10; m3.phong.toon = true;
    
    var materials = [m0, m1, m2, m3];

    var model = app.gfx.models.file('WaltHead');
    
    var sphere = app.gfx.shape(new two.Sphere(0.5));
    var big_sphere = app.gfx.shape(new two.Sphere(0.5 * 6.66));

    var transparent = Math.round(Math.random() * 3);
    materials[transparent].alpha.alpha.value = 0.9;

    //transparent : tIndex == = index ? true : false,
    //mtl.uniforms['opacity'].value = tIndex == = index ? 0.9 : 1;
    //mtl.side = tIndex == = index ? THREE.FrontSide : THREE.DoubleSide;

    this.lights = [];

    for(var i = 0; i < materials.length; ++i)
    {
        var material = materials[i];
        var position = new two.vec3(Math.sin(i * Math.PI / 2) * radius, 0.0, Math.cos(i * Math.PI / 2) * radius);
        var rotation = new two.quat(new two.vec3(0.0, i * Math.PI / 2, 0.0));

        var n = scene.nodes().add(new two.Node3(position, rotation));

        var m = scene.nodes().add(new two.Node3());
        m.derive(n, new two.vec3(0.0, -37.0, 0.0));
        
        var it = scene.items().add(new two.Item(m, model, 0, material));

        for(var j = 0; j < 8; j++)
        {
            var color = two.hsl(Math.random(), 1.0, 0.5);

            var ml = app.gfx.materials.create('light' + material.name + j.toString()); 
            ml.program = solid; ml.solid.colour.value = color;

            var mla = app.gfx.materials.create('lightalpha' + material.name + j.toString());
            mla.program = solid; mla.solid.colour.value = color; mla.alpha.alpha.value = 0.033;

            var l = scene.nodes().add(new two.Node3());
            var i0 = scene.items().add(new two.Item(l, sphere, 0, ml)); // MaterialSolid(color)));
        
            //var i1 = scene.items().add(new two.Item(l, big_sphere, 0, ma)); // MaterialSolid(color), MaterialAlpha(0.033));
            //l.children[1].scale.set(6.66, 6.66, 6.66);

            var light = scene.lights().add(new two.Light(l, two.LightType.Point, false, color, 1.0, radius));
            
            var l = {
                parent: n,
                node: l,
                color: color,
                radius: radius,
                decay: 1.0,
                sy: Math.random(),
                sr: Math.random(),
                sc: Math.random(),
                py: Math.random() * Math.PI,
                pr: Math.random() * Math.PI,
                pc: Math.random() * Math.PI,
                dir: Math.random() > 0.5 ? 1.0 : -1.0
            };
            
            this.lights.push(l);
        }
    }
}

var time = app.gfx.time;

for(var i = 0; i < this.lights.length; ++i)
{
    var l = this.lights[i];
    
    var r = 0.8 + 0.2 * Math.sin(l.pr + (0.6 + 0.3 * l.sr) * time);
    var x = (Math.sin(l.pc + (0.8 + 0.2 * l.sc) * time * l.dir)) * r * radius;
    var z = (Math.cos(l.pc + (0.8 + 0.2 * l.sc) * time * l.dir)) * r * radius;
    var y = Math.sin(l.py + (0.8 + 0.2 * l.sy) * time) * r * 32.0;
    
    l.node.derive(l.parent, new two.vec3(x, y, z));
}
